### PR TITLE
Refactoring: Switch from passing compilation as a parameter to using new ConversionsBase.Compilation property instead.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 1]
```diff
@@ -1101,32 +1107,38 @@ private Conversion ClassifyImplicitBuiltInConversionFromExpression(BoundExpressi
...
             // byte[], Span<byte>, and ReadOnlySpan<byte>.
-            ConstantValue constantValue = sourceExpression.ConstantValue;
+            ConstantValue? constantValue = sourceExpression.ConstantValue;
 
             TypeSymbol destinationOriginalDefinition = destination.OriginalDefinition;
             if (constantValue is ({ IsString: true } or { IsNull: true }) &&
                 sourceExpression.Type?.SpecialType == SpecialType.System_String &&
                 (destination is ArrayTypeSymbol { IsSZArray: true, ElementType.SpecialType: SpecialType.System_Byte } || // byte[]
                  (destinationOriginalDefinition.TypeKind == TypeKind.Struct && destinationOriginalDefinition.IsRefLikeType &&
+                  compilation is not null &&
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:1126]: >compilation is not null && [](http://example.com/codeflow?start=18&length=26)

Could we simplify by asserting `Debug.Assert(compilation is not null);` at the start of the method? (Is there a caller that uses this with `Compilation == null` and doesn't hit the assert at the end of the method?)

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:1126]: > Could we simplify by asserting ```Debug.Assert(compilation is not null);``` at the start of the method? (Is there a caller that uses this with ```Compilation == null``` and doesn't hit the assert at the end of the method?)

Technically it is possible to not reach the assert at the end even if the compilation is null. Given that the value comes from a property now, I am more comfortable with avoiding a possible crash. An assert is just as good as the tests that we have and will not protect us in a case of a test hole.

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Sorry for the spam, I accidentally commented on the wrong PR here.

**Comment by AlekseyTs** [comment]
This commit was directly incorporated into a merge PR between branches.

**Pull request closed without merging**